### PR TITLE
NEW: PEDS method, Heatmap VIZ and Associated Unit tests

### PR DESCRIPTION
**Pull request by cherman2**
Added Proportional Engraftment of Donor Strains methods, and Heatmap for tracking PEDS over time per sample.

### CODE REVIEW

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: Should the helper methods below be removed, or were you saving them for later use?

**Review by lizgehret — CHANGES_REQUESTED** [review]
Great work @cherman2! Overall I think this looks good - small comment inline, and I also think it could be nice to add a usage example or two for your method. I'm happy to help you with that as well!

File: q2_fmt/_peds.py
Comment (Oddant1) [inline q2_fmt/_peds.py]: Is there any way we can make what we're excepting a bit more specific (goes for anywhere it just says `except Exception`)? The errors you reraise seem to be related to specific failure states so unless we can be really confident that's the only thing that could raise an exception in that line of code (it probably isn't) we should probably make sure the original exception is what we think it is. Also, you can do something like this
 
```
except Exception as e:
    raise KeyError('Oh no bad') from e
```
And now your Keyerror will include information about the other exception in the traceback which can be helpful sometimes.

File: q2_fmt/_peds.py
Comment (Oddant1) [inline q2_fmt/_peds.py]: An not and?

File: q2_fmt/_peds.py
Comment (Oddant1) [inline q2_fmt/_peds.py]: An not and?

File: q2_fmt/_peds.py
Comment (Oddant1) [inline q2_fmt/_peds.py]: Any reason this has parens around it?

File: q2_fmt/_peds.py
Comment (Oddant1) [inline q2_fmt/_peds.py]: Again an not and? Looks like copy-paste biting.

File: q2_fmt/_examples.py
Comment (lizgehret) [inline q2_fmt/_examples.py:146]: ❤️

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: Should 'donor' be capital case like the rest of your attrs?

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]: This is probably a matter of style/preference over anything else - but I'm seeing some inconsistencies in your variable assignment names from snake case to camel case (`variable_example_a` vs. `variableExampleA`). I think it might look nicer to have everything snake case (since that's the typical casing for variable assignment across the framework) - but like I said before, it's more of a style/preference thing than an actual issue with your code functionality or readability. @ebolyen do you think it matters one way or the other?

File: q2_fmt/plugin_setup.py
Comment (lizgehret) [inline q2_fmt/plugin_setup.py]:
```suggestion
                       ' `table` should be grouped by. This column'
```

File: q2_fmt/plugin_setup.py
Comment (lizgehret) [inline q2_fmt/plugin_setup.py]:
```suggestion
                                    ' Default behavior is to raise an error.',
```

File: q2_fmt/plugin_setup.py
Comment (lizgehret) [inline q2_fmt/plugin_setup.py:216]:
```suggestion
                      ' also contains the numerator and denominator for'
```

File: q2_fmt/plugin_setup.py
Comment (lizgehret) [inline q2_fmt/plugin_setup.py]: Should PEDs be in all caps here?

File: q2_fmt/plugin_setup.py
Comment (lizgehret) [inline q2_fmt/plugin_setup.py]: Same as above, should PEDs be in all caps?

Comment (lizgehret) [inline q2_fmt/plugin_setup.py]:
```suggestion
                      ' peds calulations. May also contain subject IDs'
```

**Review by lizgehret — COMMENTED** [review]
Nice work @cherman2, these usage examples look great! Aside from @Oddant1's comments, I only had a few syntax/grammar tweaks to add on my end.

**Review by Oddant1 — APPROVED** [review]
Looks good to me.

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
        'description': 'Sample IDs'
```

File: q2_fmt/_peds.py
Comment (lizgehret) [inline q2_fmt/_peds.py]:
```suggestion
        'description': 'Donor'
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_get_donor(self):
```
I would remove the 'check' portion of this test name since (from my understanding) it just means that you are confirming the correct donor is being pulled from your method.

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_get_subject(self):
```
Same suggestion as `test_get_donor`

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_timepoint(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_no_donors(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_incomplete_timepoints(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_incomplete_timepoints_with_flag(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_incorrect_reference_column_name(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_incorrect_group_column_name(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_incorrect_subject_column_name(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_no_feature_overlap(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_feature_overlap(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_peds_calc(self):
```

File: q2_fmt/tests/test_engraftment.py
Comment (lizgehret) [inline q2_fmt/tests/test_engraftment.py]:
```suggestion
    def test_no_feature_in_donor(self):
```

**Review by lizgehret — CHANGES_REQUESTED** [review]
Overall this looks great @cherman2! Some small suggestions inline.

**Review by lizgehret — APPROVED** [review]
LGTM!

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This one makes more sense as a `ValueError`

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This one is also a `ValueError`

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: If you avoid the `.to_frame()` you can leave the whole things as a series which will make downstream steps easier I think. Right now it looks like you have to carry `subject_column` around everywhere for a dataframe with only one column.

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: It seems there may be a corollary of `drop_incomplete_timepoints` which is mutually exclusive to this parameter, but let you drop dud timepoints.

If you wanted such a feature, it would make sense as a separate PR after this one is merged.

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This guard misses the following data:

There is still the case where there are more than one observations per-subject per-timepoint, which would mean that the following data could slip through:

| id | subject | time|
|---|---|---|
|a|s1|t1|
|b|s1|t1|
|c|s1|t1|
|d|s2|t1|
|e|s2|t2|
|f|s2|t3|

We see that the counts are:

|s1|s2|
|---|---|
|3|3|

Which matches the number of unique timepoints

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This idea is better caught by ensuring subjects are unique within timepoints. As the only way to hit this branch would be if the data was like the above table, but with extra repeated measures for a subject in a timepoint, which means that this error message isn't really the core of the issue.

Otherwise, if subjects were unique within a timepoint, I think it would be impossible for a subject to have more observations than the number of timepoints, as they are defined by those samples which have timepoints.

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This should also be a `ValueError`

**Review by ebolyen — CHANGES_REQUESTED** [review]
Really solid work so far @cherman2!

I've got a few changes requested in the validation step, and haven't had a chance to inspect the rest yet, but from a user-level it's quite straight-forward to use.

File: q2_fmt/_peds.py
Comment (cherman2) [reply q2_fmt/_peds.py]: added this as an issue!

**Review by lizgehret — APPROVED** [review]

**Comment by cherman2** [comment]
This PR currently has a bug that I am working on

File: q2_fmt/tests/data/peds_dist/dataresource.json
Comment (lizgehret) [inline q2_fmt/tests/data/peds_dist/dataresource.json]:
```suggestion
  "format": "ndjson",
  "path": "data.ndjson"
```

**Review by lizgehret — COMMENTED** [review]
frictionless is dumb, sorry - this wasn't your fault!

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: This doesn't need the .loc[] as you are indexing with a boolean vector.

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: Stray print

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: Same idea here, but it would also be best to use `!=` instead of `<` (on the off chance the logic above is wrong or becomes wrong in the future)

File: q2_fmt/_peds.py
Comment (ebolyen) [inline q2_fmt/_peds.py]: You can save some time by doing the lookup just once like:

`sample_row = metadata.loc[sample]`

**Review by ebolyen — APPROVED** [review]
Looks good, there's a few small things that you might consider changing, but none of these block this PR.

(There may also be a good way to refactor this in the future with feature-peds using numpy and masks over the whole table)

File: q2_fmt/_peds.py
Comment (cherman2) [inline q2_fmt/_peds.py]:
```suggestion
```

File: q2_fmt/_peds.py
Comment (cherman2) [inline q2_fmt/_peds.py]:
```suggestion
                                        != num_timepoints].index).to_list()
```

**Pull request merged**
